Makefile

```diff
@@ -1,17 +1,90 @@
-all: push
+# Copyright 2016 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
 
-# 0.0 shouldn't clobber any released builds
-TAG = 0.0
-PREFIX = gcr.io/google_containers/git-sync
+.PHONY: all push push-legacy container clean
 
-binary: main.go
-	CGO_ENABLED=0 GOOS=linux godep go build -a -installsuffix cgo -ldflags '-w' -o git-sync
+REGISTRY ?= gcr.io/google_containers
+IMAGE = $(REGISTRY)/git-sync-$(ARCH)
+LEGACY_AMD64_IMAGE = $(REGISTRY)/git-sync
 
-container: binary
-	docker build -t $(PREFIX):$(TAG) .
+TAG = 1.0
 
-push: container
-	gcloud docker push $(PREFIX):$(TAG)
+# Architectures supported: amd64, arm, arm64 and ppc64le
+ARCH ?= amd64
+
+# TODO: get a base image for non-x86 archs
+#       arm arm64 ppc64le
+ALL_ARCH = amd64
+
+KUBE_CROSS_IMAGE ?= gcr.io/google_containers/kube-cross
+KUBE_CROSS_VERSION ?= v1.6.3-2
+
+GO_PKG = k8s.io/git-sync
+BIN = git-sync
+
+# If you want to build all containers, see the 'all-container' rule.
+# If you want to build AND push all containers, see the 'all-push' rule.
+all: all-build
+
+sub-container-%:
+	$(MAKE) ARCH=$* container
+
+sub-push-%:
+	$(MAKE) ARCH=$* push
+
+all-build: $(addprefix bin/$(BIN)-,$(ALL_ARCH))
+
+all-container: $(addprefix sub-container-,$(ALL_ARCH))
+
+all-push: $(addprefix sub-push-,$(ALL_ARCH))
+
+build: bin/$(BIN)-$(ARCH)
+
+bin/$(BIN)-$(ARCH): FORCE
+	mkdir -p bin
+	docker run                                      \
+	    -u $$(id -u):$$(id -g)                      \
+	    -v $$(pwd):/go/src/$(GO_PKG)                \
+	    $(KUBE_CROSS_IMAGE):$(KUBE_CROSS_VERSION)   \
+	    /bin/bash -c "                              \
+	        cd /go/src/$(GO_PKG) &&                 \
+	        CGO_ENABLED=0 go build                  \
+	        -installsuffix cgo                      \
+	        -ldflags '-w'                           \
+	        -o $@"
+
+container: .container-$(ARCH)
+.container-$(ARCH): bin/$(BIN)-$(ARCH)
+	docker build -t $(IMAGE):$(TAG) --build-arg ARCH=$(ARCH) .
+ifeq ($(ARCH),amd64)
+	docker tag -f $(IMAGE):$(TAG) $(LEGACY_AMD64_IMAGE):$(TAG)
```

Comment:
What is the replacement?  rmi and rebuild?

On Wed, Sep 14, 2016 at 12:08 AM, Brian Gosnell <notifications@github.com>
wrote:

> In Makefile
> <https://github.com/kubernetes/git-sync/pull/2#discussion_r78695936>:
>
> > +	docker run                                      \
> > +	    -u $$(id -u):$$(id -g)                      \
> > +	    -v $$(pwd):/go/src/$(GO_PKG)                \
> > +	    $(KUBE_CROSS_IMAGE):$(KUBE_CROSS_VERSION)   \
> > +	    /bin/bash -c "                              \
> > +	        cd /go/src/$(GO_PKG) &&                 \
> > +	        CGO_ENABLED=0 go build                  \
> > +	        -installsuffix cgo                      \
> > +	        -ldflags '-w'                           \
> > +	        -o $@"
> > +
> > +container: .container-$(ARCH)
> > +.container-$(ARCH): bin/$(BIN)-$(ARCH)
> > +	docker build -t $(IMAGE):$(TAG) --build-arg ARCH=$(ARCH) .
> > +ifeq ($(ARCH),amd64)
> > +	docker tag -f $(IMAGE):$(TAG) $(LEGACY_AMD64_IMAGE):$(TAG)
>
> tag -f was removed in docker 1.12.0. This causes make to fail.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/git-sync/pull/2/files/b900af9de0dfbc4f4f3aa44d97dac41e70ddf260#r78695936>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVL1ugEfYQm5-qY4SN82h6DpeQhs3ks5qp52LgaJpZM4Jssne>
> .
>